### PR TITLE
feat(grpc): adding is pruned and pruning height in blockchain info API

### LINKED ISSUE #1418: Adding Prune information to BlockchainInfo API
Add two properties to BlockchainInfo API:
1- Is Pruned
2- PruningHeight

### PR DESCRIPTION
**Pull request by kehiy**
## Related issue

- Fixes #1418

### CODE REVIEW

File: state/state.go [hunk 1]
```diff
@@ -762,3 +762,21 @@ func (st *state) AllPendingTxs() []*tx.Tx {
 
 	return st.txPool.AllPendingTxs()
 }
+
+func (st *state) IsPruned() bool {
+	st.lk.RLock()
+	defer st.lk.RUnlock()
```
Comment (Ja7ad) [inline state/state.go:768]: `IsPruned` is only set when the node is running and does not have any writes during application execution. Therefore, there is no need to lock it here.

File: store/store.go [hunk 2]
```diff
@@ -487,3 +487,7 @@ func (s *store) pruneBlock(blockHeight uint32) (bool, error) {
 
 	return true, nil
 }
+
+func (s *store) RetentionBlocks() uint32 {
+	return s.config.RetentionDays
```
Comment (Ja7ad) [inline store/store.go:492]: This is retention days not block. please change to:

```go
s.config.RetentionBlocks()
```

**Review by Ja7ad — CHANGES_REQUESTED** [review]

**Comment by kehiy** [comment]
@Ja7ad Hi, change requests fixed, please review.

File: state/state.go [hunk 3]
```diff
@@ -762,3 +762,18 @@ func (st *state) AllPendingTxs() []*tx.Tx {
 
 	return st.txPool.AllPendingTxs()
 }
+
+func (st *state) IsPruned() bool {
+	return st.store.IsPruned()
```
Comment (b00f) [inline state/state.go:767]: The Public Methods should be guarded by lock.

File: store/mock.go [hunk 4]
```diff
@@ -297,3 +297,7 @@ func (*MockStore) Prune(_ func(_ bool, _ uint32) bool) error {
 func (*MockStore) IsPruned() bool {
 	return false
 }
+
+func (*MockStore) RetentionBlocks() uint32 {
```
Comment (b00f) [inline store/mock.go:301]: Do we need it?

File: store/store.go [hunk 5]
```diff
@@ -487,3 +487,7 @@ func (s *store) pruneBlock(blockHeight uint32) (bool, error) {
 
 	return true, nil
 }
+
+func (s *store) RetentionBlocks() uint32 {
```
Comment (b00f) [inline store/store.go:491]: do we need it?

File: www/grpc/proto/network.proto [hunk 6]
```diff
@@ -71,10 +61,14 @@ message GetNodeInfoResponse {
   repeated string local_addrs = 8;
   // List of protocols supported by the node.
   repeated string protocols = 9;
+  // Type of node pruned or full.
+  bool is_pruned = 10;
```
Comment (b00f) [inline www/grpc/proto/network.proto:65]: To be backward compatible please don't change the field numbers.

File: www/grpc/proto/network.proto [hunk 7]
```diff
@@ -120,3 +114,13 @@ message PeerInfo {
   // Completed download sessions with the peer.
   int32 completed_sessions = 20;
 }
+
+// Response message containing information about the node's connections.
```
Comment (b00f) [inline www/grpc/proto/network.proto:118]:
```suggestion
// ConnectionInfo contains information about the node's connections.
```

**Review by Ja7ad — APPROVED** [review]

**Review by b00f — COMMENTED** [review]
To be compatible with Bitcoin RPC, let's add these information to the blockchain-info. Check here: https://developer.bitcoin.org/reference/rpc/getblockchaininfo.html

File: www/grpc/proto/network.proto [hunk 8]
```diff
@@ -71,10 +61,14 @@ message GetNodeInfoResponse {
   repeated string local_addrs = 8;
   // List of protocols supported by the node.
   repeated string protocols = 9;
+  // Type of node pruned or full.
```
Comment (b00f) [inline www/grpc/proto/network.proto:64]:
```suggestion
  // If the blocks are subject to pruning.
```

File: www/grpc/proto/network.proto [hunk 9]
```diff
@@ -71,10 +61,14 @@ message GetNodeInfoResponse {
...
+  bool is_pruned = 10;
+  // The height of pruning.
```
Comment (b00f) [inline www/grpc/proto/network.proto:66]:
```suggestion
  // Lowest-height block stored (only present if pruning is enabled)
```

**Review by b00f — CHANGES_REQUESTED** [review]

File: state/state.go (hunk 3 above)
Comment (kehiy) [reply state/state.go:767]: Check: https://github.com/pactus-project/pactus/pull/1420#discussion_r1680422259

File: store/store.go (hunk 5 above)
Comment (kehiy) [reply store/store.go:491]: Yes, we can't access it from state otherwise.

File: www/grpc/proto/network.proto (hunk 6 above)
Comment (kehiy) [reply www/grpc/proto/network.proto:65]: We didn't changed fields. we just used some unused numbers.

Comment (kehiy) [reply www/grpc/proto/network.proto:65]: There was no field for 10 and 11 before this.

**Comment by kehiy** [comment]
@b00f Also, why we need to be compatible with Bitcoin RPC? We are not a fork or L2.

Also no bitcoin application uses Pactus. so whats the reason to be compatible?

File: store/interface.go [hunk 10]
```diff
@@ -99,6 +99,7 @@ type Reader interface {
 	LastCertificate() *certificate.BlockCertificate
 	IsBanned(addr crypto.Address) bool
 	IsPruned() bool
+	RetentionBlocks() uint32
```
Comment (b00f) [inline store/interface.go:102]: Why don't we export "Pruningheight" instead on "RetentionBlocks"?

File: www/grpc/proto/network.proto [hunk 11]
```diff
@@ -71,10 +61,14 @@ message GetNodeInfoResponse {
   repeated string local_addrs = 8;
   // List of protocols supported by the node.
   repeated string protocols = 9;
+  // If the blocks are subject to pruning.
+  bool is_pruned = 10;
```
Comment (b00f) [inline www/grpc/proto/network.proto:65]: Please moe it to BlockchAainInfo API

**Review by b00f — CHANGES_REQUESTED** [review]

**Comment by b00f** [comment]
I think pruning information fits better in BlockchainInfo rather than NodeInfo. Regarding Bitcoin compatibility, many developers are familiar with Bitcoin, so we should create something that looks familiar and is better organized.

File: store/interface.go (hunk 10 above)
Comment (kehiy) [reply store/interface.go:102]: Because the internal function that we call have the same name. why change it?

**Comment by kehiy** [comment]
@b00f Blokchain info contains the chain info which is shared, node info contains the info about one node which is not shared between all nodes.

File: store/interface.go (hunk 10 above)
Comment (b00f) [reply store/interface.go:102]: RetentionBlocks is an internal configuration. We better export something that has more general information.

File: www/grpc/proto/network.proto [hunk 12]
```diff
@@ -71,10 +61,14 @@ message GetNodeInfoResponse {
...
+  // Lowest-height block stored (only present if pruning is enabled)
+  int32 pruning_height = 11;
   // Clock offset of the node.
-  double clock_offset = 13;
```
Comment (b00f) [inline www/grpc/proto/network.proto:75]: @kehiy don't change the old numbers. Keep them the same for backward compatibility. I know we changed them in the past, but now more users are using Pactus, and we can't ask them to upgrade their applications all the time.

Comment (kehiy) [reply www/grpc/proto/network.proto:75]: @b00f Fixed, now we didn't changes the numbers, only added some of them. anything must be backward compatible now.

**Comment by kehiy** [comment]
>  Note: we decided that to move these info to blockchain info instead of node info since it will be mostly used there, using this info you can find you whether this node can provide info about old blocks or not. In this way caller can check this even if node info API is disabled.

> Note2: we moved pruning height logic to store from state.

**Review by b00f — APPROVED** [review]

**Pull request merged**